### PR TITLE
applications: machine_learning: Add documentation

### PR DESCRIPTION
**Pull request by MarekPieta**
Change adds basic documentation for the application.

Jira: NCSDK-8466

### CODE REVIEW

File: applications/machine_learning/README.rst [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+===============
+
+The application uses :ref:`Edge Impulse's data forwarder` protocol to forward data to :ref:`Edge Impulse studio` over :ref:`zephyr:uart_api` or :ref:`nus_service_readme`.
+
+By default, the following transports are used:
+
+* **Thingy:52** uses NUS.
+* **nRF52840 Development Kit** uses UART.
+
+Machine learning model
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:47]: The intro mentions LED usage, but there is no mention of LEDs in this section.

File: applications/machine_learning/README.rst [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+======================
+
+The machine learning model is handled using :ref:`ei_wrapper` available in |NCS|.
+
+By default, the application uses pretrained models deployed in :ref:`Edge Impulse studio`:
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:52]: What does `pretrained` mean?

File: applications/machine_learning/README.rst [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+
+By default, the following transports are used:
+
+* **Thingy:52** uses NUS.
+* **nRF52840 Development Kit** uses UART.
+
+Machine learning model
+======================
+
+The machine learning model is handled using :ref:`ei_wrapper` available in |NCS|.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:50]: `in the |NCS|` (global edit)

File: applications/machine_learning/README.rst [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+
+By default, the application uses pretrained models deployed in :ref:`Edge Impulse studio`:
+
+* On **Thingy:52** - the model uses built-in accelerometer data to recognize the following gestures:
+
+  * ``idle`` - the device is placed on a flat surface.
+  * ``updown`` - the device is moved in updown direction.
+  * ``rotate`` - the device is rotated.
+  * ``tap`` - the device is tapped while placed on a flat surface.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:59]: Why are the gesture names in code formatting? I'd add an explanation where I can find these code names in the app/EI app.

File: applications/machine_learning/README.rst [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+  The unknown gestures, like for example shaking the device, are recognized as anomaly.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:61]: Is anomaly a gesture type, like ``anomaly``? If so, I'd add it to the list.

File: applications/machine_learning/README.rst [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+
+* On **nRF52840 Development Kit**  - the model uses simulated sensor data to recognize the following simulated wave types:
+
+  * ``sine``
+  * ``triangle``
+  * ``idle``
+
+  The ``square`` wave signal can also be generated by the simulated sensor.
+  This signal is unknown to the machine learning model and therefore it's marked as an anomaly.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:70]:
```suggestion
  This signal is unknown to the machine learning model and therefore it is marked as anomaly.
```

File: applications/machine_learning/README.rst [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+* **nRF52840 Development Kit** uses UART.
+
+Machine learning model
+======================
+
+The machine learning model is handled using :ref:`ei_wrapper` available in |NCS|.
+
+By default, the application uses pretrained models deployed in :ref:`Edge Impulse studio`:
+
+* On **Thingy:52** - the model uses built-in accelerometer data to recognize the following gestures:
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:54]: We use bold only for PCB elements. No formatting should be applied around board names.

File: applications/machine_learning/README.rst [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+#. Press ``Button 3`` to again change generated acceleration signal.
+   The ``square`` wave (3 blinks) is displayed only by the ``LED2``.
+   This signal is marked as anomaly by the machine learning model and ``LED1`` starts breathing.
+#. Press and hold ``Button 1`` the for time longer than 5 seconds to switch to data forwarding mode.
+   After the mode is switched, ``LED1`` starts to blink rapidly.
+   Connect to a kit with a terminal emulator (for example, PuTTY).
+   See :ref:`putty` for the required settings.
+#. Observe sensor readouts represented as comma-separated values.
+   Every line represents a single sensor readout.
+#. Turn off putty to ensure that only one program will access data on UART.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:361]: PuTTY

**Review by greg-fer — CHANGES_REQUESTED** [review]
Some early comments.

File: doc/nrf/ug_edge_impulse.rst [hunk 9]
```diff
@@ -43,7 +43,7 @@ This tutorial will guide you through the following steps:
 
    .. note::
      You can use one of the development boards supported directly by Edge Impulse or your mobile phone to collect the data.
-     You can also modify the :ref:`ei_data_forwarder_sample` sample and use it to forward data from a sensor that is connected to any board available in |NCS|.
+     You can also modify the :ref:`ei_data_forwarder_sample` sample or :ref:`nrf_machine_learning_app` application and use it to forward data from a sensor that is connected to any board available in |NCS|.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply doc/nrf/ug_edge_impulse.rst:46]: done

File: applications/machine_learning/README.rst [hunk 10]
```diff
@@ -0,0 +1,426 @@
+.. _nrf_machine_learning_app:
+
+nRF machine learning
+####################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The nRF machine learning application is an out of the box reference design of an :ref:`embedded machine learning` using :ref:`Edge Impulse`.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/machine_learning/README.rst:10]: Ok, done

File: applications/machine_learning/README.rst [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+The application performs the following tasks:
+
+* Gathers data from sensors
+* Forwards data to Edge Impulse
+* Runs machine learning model and displays results on LEDs
+
+Overview
+********
+
+As it was mentioned before, the nRF machine learning application performs multiple tasks.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/machine_learning/README.rst:20]: removed

File: applications/machine_learning/README.rst [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+The tasks are described by the paragraphs below.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/machine_learning/README.rst:21]: ok, changed

File: applications/machine_learning/README.rst (hunk 2 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:52]: The model that was trained before on a predefined dataset to perform given tasks (e.g. for our gesture classification task)

File: applications/machine_learning/README.rst (hunk 3 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:50]: done

File: applications/machine_learning/README.rst (hunk 1 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:47]: ok, I will mention it in overview

File: applications/machine_learning/README.rst (hunk 4 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:59]: These names are specific to used machine learning model. In Edge Impulse studio you define labels that should be assigned to given gestures. I will add information about this.
In application source code this names are used by the `ml_result_event` and `led_state_def.h` file. Still, the application only refers to labels (strings) that are defined by the machine learning model.

File: applications/machine_learning/README.rst (hunk 5 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:61]: Anomaly in our case is a separate thing.
The neural network reports assigned label and value related to the label (as classification results).
Apart from the neural network that performs classification task, the application uses K-means anomaly detection algorithm to detect anomalies (the input data that is different than all of the data used to train the model). If reported anomaly score is too high, then classification result reported by the neural network is ignored and the result is marked as anomaly.

File: applications/machine_learning/README.rst (hunk 6 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:70]: done

File: applications/machine_learning/README.rst (hunk 8 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:361]: fixed

File: applications/machine_learning/README.rst (hunk 7 above)
Comment (MarekPieta) [reply applications/machine_learning/README.rst:54]: ok, fixed

File: applications/machine_learning/README.rst [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+`Module event tables`_ for some modules include extensive lists of source and sink modules.
+These are valid for events that have many listeners or sources, and are gathered on the :ref:`nrf_machine_learning_app_event_rel_modules` subpage.
+
+.. toctree::
+   :maxdepth: 1
+   :caption: Subpages:
+
+   doc/event_rel_modules.rst
+   doc/main.rst
+   doc/led_state.rst
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:483]: These module pages need more details.

File: applications/machine_learning/README.rst [hunk 14]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+The nRF Machine learning application uses also the following dedicated application modules:
+
+* ``ei_data_forwarder_bt_nus``
+  The module forwards the sensor readouts over NUS to a Bluetooth connected central.
+  The sensor data is forwarded only if the connection is secured and connection interval is within the limit defined by :option:`CONFIG_BT_PERIPHERAL_PREF_MAX_INT` and :option:`CONFIG_BT_PERIPHERAL_PREF_MAX_INT`.
+* ``ei_data_forwarder_uart``
+  The module forwards the sensor readouts over UART.
+* ``led_state``
+  The module displays the application state using LEDs.
+  The LED effects used to display state of the data forwarding, machine learning results and simulated signal are defined in :file:`led_state_def.h` file located in the application configuration directory.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:453]:
```suggestion
  The LED effects used to display the state of data forwarding, the machine learning results, and the state of the simulated signal are defined in :file:`led_state_def.h` file located in the application configuration directory.
```

File: applications/machine_learning/README.rst [hunk 15]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+  Common LED effects are used to represent the machine learning results and the simulated sensor signal.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:454]:
```suggestion
  The common LED effects are used to represent the machine learning results and the simulated sensor signal.
```

File: applications/machine_learning/README.rst [hunk 16]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+* ``ml_runner``
+  The module uses :ref:`ei_wrapper` API to control running the machine learning model.
+  The module provides the prediction results using :c:struct:`ml_result_event`.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:457]:
```suggestion
  It provides the prediction results using :c:struct:`ml_result_event`.
```

File: applications/machine_learning/README.rst [hunk 17]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+  The module runs the machine learning model and provides results only if there is an active subsriber.
+  An application modules can inform that it is actively listening for results using an :c:struct:`ml_result_signin_event`.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:459]:
```suggestion
  An application module can inform that it is actively listening for results using :c:struct:`ml_result_signin_event`.
```

File: applications/machine_learning/README.rst [hunk 18]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+* ``ml_state``
+  The module controls switching between running the machine learning model and forwarding the data.
+  The change is triggered by long press of the button defined in the module's configuration.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:462]:
```suggestion
  The change is triggered by a long press of the button defined in the module's configuration.
```

File: applications/machine_learning/README.rst [hunk 19]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+* ``sensor_sim_ctrl``
+  The module controls parameters of the generated simulated sensor signal.
+  The module switches between predefined sets of parameters for the simulated signal.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:465]:
```suggestion
  It switches between predefined sets of parameters for the simulated signal.
```

File: applications/machine_learning/README.rst [hunk 20]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+  Parameters of the generated signals are defined by the :file:`sensor_sim_ctrl_def.h` file located in the application configuration directory.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:466]:
```suggestion
  The parameters of the generated signals are defined by the :file:`sensor_sim_ctrl_def.h` file located in the application configuration directory.
```

File: applications/machine_learning/README.rst [hunk 21]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+
+.. warning::
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:468]:
```suggestion
.. note::
```
We use warnings only for application-breaking cases.

File: applications/machine_learning/README.rst [hunk 22]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+   The ``ei_data_forwarder_bt_nus`` and ``ei_data_forwarder_uart`` stop forwarding the sensor readouts if they receive a :c:struct:`sensor_event` that cannot be forwarded and needs to be dropped.
+   This could happen, for example, if the selected sensor sampling frequency is too high for used implementation of the Edge Impulse data forwarder.
```
Comment (greg-fer) [inline applications/machine_learning/README.rst:470]:
```suggestion
   This could happen, for example, if the selected sensor sampling frequency is too high for the used implementation of the Edge Impulse data forwarder.
```

**Review by greg-fer — CHANGES_REQUESTED** [review]
Some minor fixes.

File: applications/machine_learning/README.rst [hunk 23]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
-These are valid for events that have many listeners or sources, and are gathered on the :ref:`nrf_machine_learning_app_event_rel_modules` subpage.
-
-.. toctree::
-   :maxdepth: 1
-   :caption: Subpages:
 
-   doc/event_rel_modules.rst
-   doc/main.rst
-   doc/led_state.rst
+The nRF Machine Learning application uses the following modules available in CAF:
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/machine_learning/README.rst:433]: Ok, added your change

File: applications/machine_learning/README.rst [hunk 24]
```diff
@@ -468,16 +429,43 @@ Application internal modules
...
+  The change is triggered by long press of the button defined in the module's configuration.
+* ``sensor_sim_ctrl``
+  The module controls parameters of the generated simulated sensor signal.
+  The module switches between predefined sets of parameters for the simulated signal.
+  Parameters of the generated signals are defined by the :file:`sensor_sim_ctrl_def.h` file located in the application configuration directory.
+
+.. warning::
+   The ``ei_data_forwarder_bt_nus`` and ``ei_data_forwarder_uart`` stop forwarding the sensor readouts if they receive a :c:struct:`sensor_event` that cannot be forwarded and needs to be dropped.
+   This could happen, for example, if the selected sensor sampling frequency is too high for used implementation of the Edge Impulse data forwarder.
+   Data forwaring is stopped to make sure that dropping samples will be noticed by the user.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/machine_learning/README.rst]: done

**Review by greg-fer — APPROVED** [review]
Let's get this baby in.

**Review by pdunaj — APPROVED** [review]
